### PR TITLE
Moved creation `Agent` on receiving a first heartbeat

### LINKED ISSUE #1595: Move creation of agent to state when a first heartbeat is received

### PR DESCRIPTION
**Pull request by nulls**
### What's done:
* moved creation of `Agent` to heartbeat with state = STARTING
* combined logic for creating and starting container to a single command

It closes #1595

It depends to #1630

----
Extracted refactoring:
- [x] https://github.com/saveourtool/save-cloud/pull/1626
- [x] https://github.com/saveourtool/save-cloud/pull/1627
- [x] https://github.com/saveourtool/save-cloud/pull/1628
- [x] https://github.com/saveourtool/save-cloud/pull/1642
- [x] https://github.com/saveourtool/save-cloud/pull/1643
- [x] https://github.com/saveourtool/save-cloud/pull/1650

### CODE REVIEW

File: save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/AgentService.kt [hunk 1]
```diff
@@ -57,27 +55,16 @@ class AgentService(
...
-    fun saveAgentsWithInitialStatuses(
+    fun addAgent(
         executionId: Long,
-        agents: List<AgentDto>,
-    ): Mono<EmptyResponse> = agents.toFlux()
-        .flatMap { agent ->
-            orchestratorAgentService
-                .addAgent(executionId, agent)
-                .flatMap {
-                    orchestratorAgentService.updateAgentStatus(
```
[the remark this answers is not in the document]
Comment (nulls) [reply save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/AgentService.kt:75]: yes, the status will be updated after getting a heartbeat, but agent is created after it

and it's look like we should get *not found* issue: we try to update AgentStatus before Agent is created.
will re-test (local test doesn't show this error)

**Review by 0x6675636b796f75676974687562 — APPROVED** [review]

File: save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/AgentService.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (nulls) [reply save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/AgentService.kt:75]: After testing, fixed the issues:
1. Now we call `addAgent` before updating `agentStatus` in DB
2. Now we finalize execution only if there are no active containers in `inmemory` repository

File: save-agent/src/commonMain/kotlin/com/saveourtool/save/agent/utils/HttpUtils.kt [hunk 2]
```diff
@@ -92,7 +92,7 @@ internal suspend fun HttpClient.download(url: String, body: Any?, file: Path): R
                                 it.write(bytes)
                             }
                         totalBytes.addAndGet(bytes.size.toLong())
-                        logDebugCustom("Received ${bytes.size} bytes out of ${httpResponse.contentLength()} bytes from ${httpResponse.request.url}")
+                        logDebugCustom("Received ${bytes.size} ($totalBytes) bytes out of ${httpResponse.contentLength()} bytes from ${httpResponse.request.url}")
```
Comment (nulls) [inline save-agent/src/commonMain/kotlin/com/saveourtool/save/agent/utils/HttpUtils.kt:95]: before, we get a lot of similar log message in save-agent

File: save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/AgentService.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (nulls) [reply save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/AgentService.kt:75]: > 2. Now we finalize execution only if there are no active containers in inmemory repository

This actually doesn't work in all cases, due to we run this action in async mode. So refactored `inmemory` repository to no check that repository knows about execution which is going to be clean-up

File: save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/AgentService.kt [hunk 3]
```diff
@@ -110,22 +97,20 @@ class AgentService(
...
-            .filter { finishedContainerIds ->
-                finishedContainerIds.isNotEmpty()
-                    .also { hasFinishedContainers ->
-                        if (!hasFinishedContainers) {
-                            log.debug { "Agents for execution $executionId are still running, so won't try to stop them" }
-                        }
-                    }
+            .filter { haveFinalStatus ->
+                if (!haveFinalStatus) {
+                    log.debug { "Agents for execution $executionId are still running, so won't try to stop them" }
```
[the remark this answers is not in the document]
Comment (nulls) [reply save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/AgentService.kt]: done

**Pull request merged**